java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -51,191 +51,211 @@
                                                            List<Class<? extends Constraint>> domains) {
     return flow(currentNode, currentVal, addToFlow, c -> false, domains);
   }
 
-  public static List<JavaFileScannerContext.Location> flow(ExplodedGraph.Node currentNode, @Nullable SymbolicValue currentVal, Predicate<Constraint> addToFlow,
+  public static Set<List<JavaFileScannerContext.Location>> flow(ExplodedGraph.Node currentNode, @Nullable SymbolicValue currentVal, Predicate<Constraint> addToFlow,
                                                            Predicate<Constraint> terminateTraversal, List<Class<? extends Constraint>> domains) {
     FlowComputation flowComputation = new FlowComputation(currentVal, addToFlow, terminateTraversal, domains);
-
+    // FIXME SONARJAVA-2049 getLastEvaluated doesn't work with relational SV
     Symbol trackSymbol = currentNode.programState.getLastEvaluated();
-    if (currentVal instanceof BinarySymbolicValue) {
-      Set<JavaFileScannerContext.Location> binSVFlow = flowComputation.flowFromBinarySV(currentNode, (BinarySymbolicValue) currentVal, trackSymbol);
-      flowComputation.flow.addAll(binSVFlow);
-    }
-    flowComputation.run(currentNode, trackSymbol);
-    return flowComputation.flow;
+    return flowComputation.run(currentNode, trackSymbol);
   }
 
-  private Set<JavaFileScannerContext.Location> flowFromBinarySV(ExplodedGraph.Node currentNode, BinarySymbolicValue binarySV, Symbol trackSymbol) {
-    Set<JavaFileScannerContext.Location> binSVFlow = new LinkedHashSet<>();
-    FlowComputation left = fork(binarySV.getLeftOp());
-    left.run(currentNode.parent(), trackSymbol);
-    binSVFlow.addAll(left.flow);
-    FlowComputation right = fork(binarySV.getRightOp());
-    right.run(currentNode.parent(), trackSymbol);
-    binSVFlow.addAll(right.flow);
-    return binSVFlow;
-  }
+  private Set<List<JavaFileScannerContext.Location>> run(final ExplodedGraph.Node node, @Nullable final Symbol trackSymbol) {
+    Set<List<JavaFileScannerContext.Location>> flows = new HashSet<>();
+    Deque<ExecutionPath> workList = new ArrayDeque<>();
 
-  private FlowComputation fork(SymbolicValue symbolicValue) {
-    return new FlowComputation(symbolicValue, addToFlow, terminateTraversal, domains);
+    node.edges().forEach(e -> workList.push(new ExecutionPath(trackSymbol).addEdge(e)));
+    while (!workList.isEmpty()) {
+      ExecutionPath path = workList.pop();
+      if (path.finished) {
+        flows.add(path.flow);
+      } else {
+        path.lastEdge.parent.edges().stream()
+          .filter(path::notVisited)
+          .map(path::addEdge)
+          .forEach(workList::push);
+      }
+    }
+    return flows;
   }
 
-  private static class NodeSymbol {
-    final ExplodedGraph.Node node;
-    final Symbol trackSymbol;
+  private class ExecutionPath {
+    final @Nullable Symbol trackSymbol;
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AModifiersOrderCheck)